File: python/table_4.py
""" Table 1 in Section 8.1 in Canay, Illanes and Velez (2023)
"""

import time
from pathlib import Path

import numpy as np
import texttable as tt
from latextable import draw_latex
from scipy.optimize import NonlinearConstraint, minimize

import ineq_functions as ineq

# Make folders for results
results_dir = Path(__file__).resolve().parent / "_results"
tables_dir = results_dir / "tables-tex"
tables_dir.mkdir(parents=True, exist_ok=True)


def load_data(name):
    file_path = Path(__file__).resolve().parent.parent / "data" / (name + ".csv")
    return np.loadtxt(file_path, delimiter=",")


dgp = {i: load_data(i) for i in ["A", "D", "J0", "Dist"]}
dgp["Dist"] = dgp["Dist"] / 1000  # Rescale the distance
dgp["W"] = dgp["D"][:, 1:]
n = dgp["A"].shape[0]

settings = {
    "Vbar": [500, 500, 1000, 1000],
    "test_stat": ["CCK", "CCK", "CCK", "CCK"],
    "cv": ["SN", "SN", "SN", "SN"],
    "alpha": 0.05,
    "IV": None,
}

sim = {
    "grid_size": 1401,
    "rng_seed": 20220826,
    "bootstrap_replications": 1000,
    "sim_name": "table_4",
    "lb": np.array(
        [
            [-40, -20, 0, -40, -20, 0, 0, 0],
            [-40, -20, -10, -40, -20, -10, 0, 0],
            [-40, -20, 0, -40, -20, 0, 0, 0],
            [-40, -20, -10, -40, -20, -10, 0, 0],
        ]
    ),
    "ub": np.array(
        [
            [100, 50, 0, 100, 50, 0, 3, 2],
            [100, 50, 10, 100, 50, 10, 3, 2],
            [100, 50, 0, 100, 50, 0, 3, 2],
            [100, 50, 10, 100, 50, 10, 3, 2],
        ]
    ),
}
sim["x0"] = np.zeros((4, 8))

results = {
    "CI_vec": [np.full((4, 2), np.nan) for i in range(8)],
    "comp_time": np.empty(4),
}

# Generate bootstrap indices
np.random.seed(sim["rng_seed"])
bootstrap_indices = np.random.randint(0, n, size=(sim["bootstrap_replications"], n))


# Define the constraint function
def restriction_function(
    theta: np.ndarray, sim_i: int, account_uncertainty: bool = False
):
    """Wrapper function for :func:`g_restriction` to be used with scipy.optimize"""
    if theta.shape == (1,):
        raise ValueError("theta must be a 1d array")
    # Return the difference between the critical value and test stat
    return -ineq.g_restriction_diff(
        theta=theta,
        W_data=dgp["W"],
        A_matrix=dgp["A"],
        J0_vec=dgp["J0"],
        Vbar=settings["Vbar"][sim_i],
        IV_matrix=settings["IV"],
        grid0="all",
        alpha=settings["alpha"],
        test0=settings["test_stat"][sim_i],
        cvalue=settings["cv"][sim_i],
        account_uncertainty=account_uncertainty,
        bootstrap_indices=bootstrap_indices,
        dist_data=dgp["Dist"],
    )


for sim_i in range(4):
    print("Simulation:", sim_i + 1)
    # Obtain the time at the beginning of the simulation
    tic = time.perf_counter()

    # Six dimensional theta confidence intervals
    for theta_index in range(6):
        # Define the nonlinear constraint
        nonlinear_constraint = {
            "type": "ineq",
            "fun": restriction_function,
            "args": (sim_i,),
        }

        if sim["lb"][sim_i, theta_index] == sim["ub"][sim_i, theta_index]:
            # If the bounds are equal, then theta is fixed
            results["CI_vec"][theta_index][sim_i, 0] = sim["lb"][sim_i, theta_index]
            results["CI_vec"][theta_index][sim_i, 1] = sim["ub"][sim_i, theta_index]
        else:
            # Call the optimization routine
            CI_lower = minimize(
                lambda x, i: x[i],
                x0=sim["x0"][sim_i, 0:6],
                args=(theta_index,),
                method="SLSQP",
                jac=lambda x, i: np.array([1 if j == i else 0 for j in range(6)]),
                bounds=[
                    (sim["lb"][sim_i, t_i], sim["ub"][sim_i, t_i]) for t_i in range(6)
                ],
                constraints=nonlinear_constraint,
                tol=1e-8,
            )

            CI_upper = minimize(
                lambda x, i: -x[i],
                sim["x0"][sim_i, 0:6],
                args=(theta_index,),
                method="SLSQP",
                jac=lambda x, i: np.array([-1 if j == i else 0 for j in range(6)]),
                bounds=[
                    (sim["lb"][sim_i, t_i], sim["ub"][sim_i, t_i]) for t_i in range(6)
                ],
                constraints=nonlinear_constraint,
                tol=1e-8,
            )

            results["CI_vec"][theta_index][sim_i, 0] = CI_lower.x[theta_index]
            results["CI_vec"][theta_index][sim_i, 1] = CI_upper.x[theta_index]

    # Two dimensional theta confidence intervals accounting for uncertainty
    for theta_index in range(2):
        # Define the nonlinear constraint (this time accounting for uncertainty)
        nonlinear_constraint = {
            "type": "ineq",
            "fun": restriction_function,
            "args": (sim_i, True),  # True: account for uncertainty
        }

        # Call the optimization routine
        CI_lower = minimize(
            lambda x, i: np.sum(x[(i * 3) : ((i + 1) * 3)]) * x[6 + i],
            x0=sim["x0"][sim_i],
            args=(theta_index,),
            method="SLSQP",
            jac=lambda x, i: np.array(
                [x[6 + i] if (i * 3) <= j < ((i + 1) * 3) else 0 for j in range(6)]
                + [
                    np.sum(x[(i * 3) : ((i + 1) * 3)]) if j == 6 + i else 0
                    for j in range(6, 8)
                ]
            ),
            bounds=[(sim["lb"][sim_i, t_i], sim["ub"][sim_i, t_i]) for t_i in range(8)],
            constraints=nonlinear_constraint,
            tol=1e-8,
        )

        CI_upper = minimize(
            lambda x, i: -np.sum(x[(i * 3) : ((i + 1) * 3)]) * x[6 + i],
            sim["x0"][sim_i],
            args=(theta_index,),
            method="SLSQP",
            jac=lambda x, i: -np.array(
                [x[6 + i] if (i * 3) <= j < ((i + 1) * 3) else 0 for j in range(6)]
                + [
                    np.sum(x[(i * 3) : ((i + 1) * 3)]) if j == 6 + i else 0
                    for j in range(6, 8)
                ]
            ),
            bounds=[(sim["lb"][sim_i, t_i], sim["ub"][sim_i, t_i]) for t_i in range(8)],
            constraints=nonlinear_constraint,
            tol=1e-8,
        )

        results["CI_vec"][theta_index + 6][sim_i, 0] = CI_lower.x[theta_index]
        results["CI_vec"][theta_index + 6][sim_i, 1] = CI_upper.x[theta_index]

    # Stop the timer
    toc = time.perf_counter()
    results["comp_time"][sim_i] = toc - tic
    print("~> time:", results["comp_time"][sim_i])

# Save results
(results_dir / sim["sim_name"]).mkdir(exist_ok=True)
for key, value in results.items():
    np.save(results_dir / sim["sim_name"] / key, value)

# Make and print the table
tableObj = tt.Texttable(0)

# tableObj.set_cols_align(["c", "c", "c", "c", "c", "c"])
# tableObj.set_cols_dtype(["t", "t", "t", "t", "t", "t"])
#
the_table = np.array(["Coca-", "Cola", "", "", "Energy", "Brands", "", ""])
the_table = np.column_stack(
    (
        the_table,
        np.array(
            [
                f"$\\theta_{{{i},{j}}}$" if j < 4 else f"$\\theta_{{{i}}}$"
                for i in range(1, 3)
                for j in range(1, 5)
            ]
        ),
    )
)

sub_table = []
for ci_theta in results["CI_vec"]:
    sub_table += [
        [
            "[" + "{:.1f}".format(x[0]) + ", " + "{:.1f}".format(x[1]) + "]"
            for x in ci_theta
        ]
    ]

# the order has the theta_1(mu) and theta_2(mu) at the end
sorted_sub_table = np.array(sub_table)[[0, 1, 2, 6, 3, 4, 5, 7], :]
the_table = np.column_stack((the_table, sorted_sub_table))
the_table = np.vstack(
    (
        ["", "Parameter", "Linear", "Quadratic", "Linear", "Quadratic"],
        the_table,
        ["Comp. Time", ""] + list(results["comp_time"]),
    )
)

tableObj.add_rows(the_table)
print(tableObj.draw())

# Save table
with open(tables_dir / (sim["sim_name"] + ".tex"), "w", encoding="utf8") as f:
    f.write(draw_latex(tableObj))

File: r/table_4.R
# Table 1 in Section 8.1 in Canay, Illanes and Velez (2023)

if (!dir.exists("_results")) {
  dir.create("_results")
}

# Import packages and functions
require(readr)
require(tictoc)
require(Rfast)
require(Rfast2)
require(nloptr)
require(xtable)
require(devtools)

# Install and import package
devtools::install("ineqfunctions", upgrade = "never")

# Import data
datasets <- c("A", "D", "J0", "Dist")
data_path <- file.path("..", "data")
dgp <- sapply(datasets, function(dataset) {
  unname(as.matrix(readr::read_csv(
    file.path(data_path, paste0(dataset, ".csv")),
    col_names = F,
    show_col_types = F
  )))
}, simplify = F)
dgp$Dist <- dgp$Dist / 1000
dgp$W <- dgp$D[, -1]

# Settings (cell arrays are used to loop over each of the four different specifications)
settings <- list(
  v_bar = c(500, 500, 1000, 1000),
  test_stat = c("CCK", "CCK", "CCK", "CCK"),
  cv = c("SN", "SN", "SN", "SN"),
  alpha = 0.05,
  iv = NULL
)

# Technical settings (lists are used to loop over the two parameters: theta1 and theta2)
sim <- list(
  sim_name = "table_4",
  lb = rbind(
    c(-40, -20, 0, -40, -20, 0, 0, 0),
    c(-40, -20, -10, -40, -20, -10, 0, 0),
    c(-40, -20, 0, -40, -20, 0, 0, 0),
    c(-40, -20, -10, -40, -20, -10, 0, 0)
  ),
  ub = rbind(
    c(100, 50, 0, 100, 50, 0, 3, 2),
    c(100, 50, 10, 100, 50, 10, 3, 2),
    c(100, 50, 0, 100, 50, 0, 3, 2),
    c(100, 50, 10, 100, 50, 10, 3, 2)
  ),
  x0 = matrix(0, nrow = 4, ncol = 8),
  opt = list(
    algorithm = "NLOPT_LD_SLSQP",
    maxeval = 1e5,
    xtol_rel = 5e-4,
    ftol_rel = 1e-8
  )
)

results <- list(
  ci_vector = lapply(1:8, function(i) {
    matrix(NA, nrow = 4, ncol = 2)
  }),
  comp_time = rep(NA, 4)
)

# Define constraint function
restriction_function <- function(theta, sim_i, theta_index, account_uncertainty) {
  restriction_terms <- ineqfunctions::g_restriction(
    theta = theta,
    w_data = dgp$W,
    a_matrix = dgp$A,
    j0_vector = dgp$J0,
    v_bar = settings$v_bar[sim_i],
    alpha = settings$alpha,
    grid0 = "all",
    test0 = settings$test_stat[sim_i],
    cvalue = settings$cv[sim_i],
    iv_matrix = settings$iv,
    account_uncertainty = account_uncertainty,
    dist_data = dgp$Dist,
  )
  restriction_terms[1] - restriction_terms[2]
}

restriction_jac <- function(theta, sim_i, theta_index, account_uncertainty) {
  nloptr::nl.grad(
    x0 = theta,
    fn = restriction_function,
    heps = 1e-6,
    sim_i = sim_i,
    theta_index = theta_index,
    account_uncertainty = account_uncertainty
  )
}

for (sim_i in 1:4) {
  tictoc::tic(paste("Simulation", sim_i))

  # Six dimensional theta confidence intervals
  for (theta_index in 1:6) {
    if (sim$lb[sim_i, theta_index] == sim$ub[sim_i, theta_index]) {
      # If the bounds are equal, then theta is fixed
      results$ci_vector[[theta_index]][sim_i, ] <- c(
        sim$lb[sim_i, theta_index],
        sim$ub[sim_i, theta_index]
      )
    } else {
      # Call the optimization routine
      ci_lower <- nloptr::nloptr(
        x0 = as.numeric(sim$x0[sim_i,1:6]),
        eval_f = function(theta, sim_i, theta_index, account_uncertainty) {
          theta[theta_index]
        },
        eval_grad_f = function(theta, sim_i, theta_index, account_uncertainty) {
          grad_f <- numeric(6)
          grad_f[theta_index] <- 1
          grad_f
        },
        lb = sim$lb[sim_i, 1:6],
        ub = sim$ub[sim_i, 1:6],
        eval_g_ineq = restriction_function,
        eval_jac_g_ineq = restriction_jac,
        opts = sim$opt,
        sim_i = sim_i,
        theta_index = theta_index,
        account_uncertainty = FALSE
      )

      ci_upper <- nloptr::nloptr(
        x0 = as.numeric(sim$x0[sim_i,1:6]),
        eval_f = function(theta, sim_i, theta_index, account_uncertainty) {
          -theta[theta_index]
        },
        eval_grad_f = function(theta, sim_i, theta_index, account_uncertainty) {
          grad_f <- numeric(6)
          grad_f[theta_index] <- -1
          grad_f
        },
        lb = as.numeric(sim$lb[sim_i, 1:6]),
        ub = as.numeric(sim$ub[sim_i, 1:6]),
        eval_g_ineq = restriction_function,
        eval_jac_g_ineq = restriction_jac,
        opts = sim$opt,
        sim_i = sim_i,
        theta_index = theta_index,
        account_uncertainty = FALSE
      )

      results$ci_vector[[theta_index]][sim_i, 1] <- ci_lower$solution[theta_index]
      results$ci_vector[[theta_index]][sim_i, 2] <- ci_upper$solution[theta_index]
    }
  }

  # Two dimensional theta confidence intervals accounting for uncertainty
  for (theta_index in 1:2) {
    # Call the optimization routine
    ci_lower <- nloptr::nloptr(
      x0 = as.numeric(sim$x0[sim_i,]),
      eval_f = function(theta, sim_i, theta_index, account_uncertainty) {
        sum(theta[((theta_index - 1) * 3 + 1) : ((theta_index) * 3)]) * theta[6 + theta_index]
      },
      eval_grad_f = function(theta, sim_i, theta_index, account_uncertainty) {
        c(
          sapply(1:6, function(j) ifelse(((theta_index - 1) * 3 + 1) <= j & j < ((theta_index - 1) * 3 + 4), theta[6 + theta_index], 0)),
          sapply(7:8, function(j) ifelse(j == (6 + theta_index), sum(theta[((theta_index - 1) * 3 + 1) : ((theta_index - 1) * 3 + 3)]), 0))
        )
      },
      lb = as.numeric(sim$lb[sim_i,]),
      ub = as.numeric(sim$ub[sim_i,]),
      eval_g_ineq = restriction_function,
      eval_jac_g_ineq = restriction_jac,
      opts = sim$opt,
      sim_i = sim_i,
      theta_index = theta_index,
      account_uncertainty = TRUE
    )

    ci_upper <- nloptr::nloptr(
      x0 = as.numeric(sim$x0[sim_i,]),
      eval_f = function(theta, sim_i, theta_index, account_uncertainty) {
        -sum(theta[((theta_index - 1) * 3 + 1) : ((theta_index) * 3)]) * theta[6 + theta_index]
      },
      eval_grad_f = function(theta, sim_i, theta_index, account_uncertainty) {
        -c(
          sapply(1:6, function(j) ifelse(((theta_index - 1) * 3 + 1) <= j & j < ((theta_index - 1) * 3 + 4), theta[6 + theta_index], 0)),
          sapply(7:8, function(j) ifelse(j == (6 + theta_index), sum(theta[((theta_index - 1) * 3 + 1) : ((theta_index - 1) * 3 + 3)]), 0))
        )
      },
      lb = as.numeric(sim$lb[sim_i,]),
      ub = as.numeric(sim$ub[sim_i,]),
      eval_g_ineq = restriction_function,
      eval_jac_g_ineq = restriction_jac,
      opts = sim$opt,
      sim_i = sim_i,
      theta_index = theta_index,
      account_uncertainty = TRUE
    )

    results$ci_vector[[theta_index + 6]][sim_i, 1] <- ci_lower$solution[theta_index]
    results$ci_vector[[theta_index + 6]][sim_i, 2] <- ci_upper$solution[theta_index]

  }

  # Stop the timer
  temp_timer <- tictoc::toc()
  results$comp_time[sim_i] <- temp_timer$toc - temp_timer$tic
}

# 3 Save results
save(results, file = file.path("_results", paste0(sim$sim_name, ".Rdata")))

### 4 Print table
table_dir <- file.path("_results", "tables-tex")
if (!dir.exists(table_dir)) {
  dir.create(table_dir)
}

the_table <- c("Coca-Cola", "", "", "", "Energy Brands", "", "", "")
the_table <- cbind(
  the_table,
  paste0(
    rep(paste0("$\\theta_{", 1:2), each = 4),
    ifelse(1:8 %% 4 == 0, "}$", paste0(",", rep(1:4, 2), "}$"))
  )
)

# Format CI as [lb, ub]
sub_table <- sapply(results$ci_vector, function(ci_theta) {
 apply(ci_theta, 1, function(x) {
   paste0("[", sprintf("%.1f", x[1]), ", ", sprintf("%.1f", x[2]), "]")
 })
})

# the order has the theta_1(mu) and theta_2(mu) at the end
sorted_sub_table <- t(sub_table)[c(1:3,7,4:6,8),]
the_table <- cbind(the_table, sorted_sub_table)
the_table <- rbind(
  the_table,
  c("Comp. Time", "", sprintf("%.2f", results$comp_time))
)
colnames(the_table) <- c(
  " ", "Parameter", "Linear", "Quadratic", "Linear", "Quadratic"
)

# Save the table
print(
  xtable(the_table, digits = 4),
  include.rownames = FALSE,
  sanitize.text.function = identity,
  file = file.path(table_dir, paste0(sim$sim_name, ".tex"))
)
